### PR TITLE
Last three tests skip on lxc.

### LINKED ISSUE #9875: Running integration tests with lxc fail
So the `exec` tests should obviously fail, some of the `cap-add` ones make sense, and the networking ipc and mac address ones should fail as well, but there are a few additional tests that are failing that I don't think should fail.
Maybe @dineshs-altiscale can clear up some of the uncertainties here.
Failing tests:
```
--- FAIL: TestPsListContainersSize (1.22s)
	docker_cli_ps_test.go:281: Expected size "2 B", got "80 B"
--- FAIL: TestPsListContainersFilterStatus (1.98s)
	docker_cli_ps_test.go:322: Expected id 0680995f62db, got 0680995f62db
		cdede23854f1 for exited filter, output: "0680995f62db\ncdede23854f1\n"

--- FAIL: TestRunVolumesMountedAsReadonly (0.95s)
	docker_cli_run_test.go:368: run should fail because volume is ro: exit code 0
--- FAIL: TestRunVolumesFromInReadonlyMode (2.08s)
	docker_cli_run_test.go:384: run should fail because volume is ro: exit code 0
--- FAIL: TestRunVolumesFromInReadWriteMode (0.03s)
	docker_cli_run_test.go:396: exit status 1
--- FAIL: TestVolumesFromGetsProperMode (0.02s)
	docker_cli_run_test.go:422: exit status 1
--- FAIL: TestRunApplyVolumesFromBeforeVolumes (0.02s)
	docker_cli_run_test.go:449: exit status 1

--- FAIL: TestRunCreateVolumeWithSymlink (1.77s)
	docker_cli_run_test.go:534: [run] err: exit status 255, exitcode: 255
--- FAIL: TestRunVolumesFromSymlinkPath (2.13s)
	docker_cli_run_test.go:578: [run] (volume) err: exit status 255, exitcode: 255

--- FAIL: TestRunEnvironment (1.13s)
	docker_cli_run_test.go:762: Wrong environment: should be 8 variables, not: ", FALSE=true, HOME=/root, HOSTNAME=testing, PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin, TRICKY=tri, TRUE=false, cky, container=lxc"

--- FAIL: TestRunCapDropInvalid (1.03s)
	docker_cli_run_test.go:898: <nil> bin
		dev
		etc
		home
		lib
		lib64
		linuxrc
		media
		mnt
		opt
		proc
		root
		run
		sbin
		sys
		tmp
		usr
		var
		

--- FAIL: TestRunCapDropALLCannotMknod (1.00s)
	docker_cli_run_test.go:938: <nil> ok
		
--- FAIL: TestRunCapDropALLAddMknodCannotMknod (1.25s)
	docker_cli_run_test.go:953: exit status 1 2014/12/31 19:58:28 setup user setgroups operation not permitted
		lxc-start: The container failed to start.
		lxc-start: Additional information can be obtained by setting the --logfile and --logpriority options.
		
--- FAIL: TestRunCapAddInvalid (1.28s)
	docker_cli_run_test.go:970: <nil> bin
		dev
		etc
		home
		lib
		lib64
		linuxrc
		media
		mnt
		opt
		proc
		root
		run
		sbin
		sys
		tmp
		usr
		var

--- FAIL: TestRunCapAddALLCanDownInterface (0.93s)
	docker_cli_run_test.go:995: exit status 255 lxc-start: unknown capability block_suspend
		lxc-start: failed to keep requested caps
		lxc-start: failed to setup the container
		lxc-start: invalid sequence number 1. expected 2
		lxc-start: failed to spawn 'f29df582a245c01a2e5a7dfa558f0647a7ac936e0f560e774b2440b9228df944'
		lxc-start: The container failed to start.
		lxc-start: Additional information can be obtained by setting the --logfile and --logpriority options.
		

--- FAIL: TestRunSysNotWritableInNonPrivilegedContainers (1.04s)
	docker_cli_run_test.go:1056: sys should not be writable in a non privileged container
[PASSED]: run - sys writable in privileged container
--- FAIL: TestRunProcNotWritableInNonPrivilegedContainers (0.96s)
	docker_cli_run_test.go:1078: proc should not be writable in a non privileged container

--- FAIL: TestRunWriteHostsFileAndNotCommit (1.12s)
	docker_cli_run_test.go:1706: diff should be empty
--- FAIL: TestRunWriteHostnameFileAndNotCommit (1.16s)
	docker_cli_run_test.go:1735: diff should be empty
--- FAIL: TestRunWriteResolvFileAndNotCommit (1.07s)
	docker_cli_run_test.go:1764: diff should be empty

--- FAIL: TestRunUnknownCommand (1.24s)
	docker_cli_run_test.go:2480: Container should not have been able to start!

--- FAIL: TestSaveDirectoryPermissions (2.67s)
	docker_cli_save_load_test.go:425: failed to find the layer with the right content listinghttps://jenkins.dockerproject.com/job/Docker%20Master%20(iffy%20builds)/label=debian-devmapper-lxc/7/console
```

Full log can be seen here: https://jenkins.dockerproject.com/job/Docker%20Master%20(iffy%20builds)/label=debian-devmapper-lxc/7/console

### PR DESCRIPTION
**Pull request by jessfraz**
Now we can scale lxc tests to all PRs.
ping @ashahab-altiscale, seriously good work getting this all done

closes #9875

### CODE REVIEW

**Comment by ashahab-altiscale** [comment]
@jfrazelle LGTM. And thanks for all the help.

**Comment by ashahab-altiscale** [comment]
@dineshs-altiscale

File: integration-cli/docker_cli_nat_test.go [hunk 1]
```diff
@@ -9,6 +9,7 @@ import (
 )
 
 func TestNetworkNat(t *testing.T) {
+	testRequires(t, NativeExecDriver)
 	testRequires(t, SameHostDaemon)
```
Comment (tiborvass) [inline integration-cli/docker_cli_nat_test.go]: @jfrazelle `testRequires(t, NativeExecDriver, SameHostDaemon)` two in one :)

File: integration-cli/docker_cli_run_test.go [hunk 2]
```diff
@@ -2887,6 +2888,7 @@ func TestContainerNetworkMode(t *testing.T) {
 }
 
 func TestRunModePidHost(t *testing.T) {
+	testRequires(t, NativeExecDriver)
```
Comment (tiborvass) [inline integration-cli/docker_cli_run_test.go]: same here

File: integration-cli/docker_cli_nat_test.go (hunk 1 above)
Comment (jfrazelle) [inline integration-cli/docker_cli_nat_test.go]: ooooo i did not know that

Comment (jfrazelle) [inline integration-cli/docker_cli_nat_test.go]: i found a few more instances where defer was before lol so im going to fix those too

**Comment by jfrazelle** [comment]
if you are wondering why so many lines changes its bc I fixed some places where defer deleteAllContainers came before the testRequires

**Comment by ahmetalpbalkan** [comment]
@jfrazelle LGTM. thanks for fixing those defers.

**Comment by jfrazelle** [comment]
Ping @shin- checkout the docker-Py tests on lxc

**Comment by jfrazelle** [comment]
also @shin- I can work on a patch

/me whips out python toolbelt

**Comment by jfrazelle** [comment]
but that should not hold up this PR ;)

File: integration-cli/requirements.go [hunk 3]
```diff
@@ -25,6 +26,13 @@ var (
 		func() bool { return supportsExec },
 		"Test requires 'docker exec' capabilities on the tested daemon.",
 	}
+	NativeExecDriver = TestRequirement{
+		func() bool {
+			driver := os.Getenv("DOCKER_EXECDRIVER")
```
Comment (tianon) [inline integration-cli/requirements.go]: I die errytime

Comment (tianon) [inline integration-cli/requirements.go]: What about remote daemons? :wink:

Comment (jfrazelle) [inline integration-cli/requirements.go]: oh man idk what to do as solution

Comment (ahmetalpbalkan) [inline integration-cli/requirements.go]: I actually double checked last night there is nothing to do about this one.

**Comment by shin-** [comment]
I can look into it. Is the fact that `ExitCode` doesn't get updated on kill/stop when using LXC the expected behavior, or am I doing something wrong?

File: integration-cli/requirements.go (hunk 3 above)
Comment (ahmetalpbalkan) [inline integration-cli/requirements.go]: I actually double checked this last night, there is nothing to do about this one.

**Comment by jfrazelle** [comment]
> I can look into it. Is the fact that ExitCode doesn't get updated on kill/stop when using LXC the expected behavior, or am I doing something wrong?

hmm @ashahab-altiscale will know best

File: integration-cli/requirements.go (hunk 3 above)
Comment (tianon) [inline integration-cli/requirements.go]: And cache it, obvs

Comment (tianon) [inline integration-cli/requirements.go]: Can't we hit the /info API endpoint to find out the exec driver?

**Comment by ashahab-altiscale** [comment]
@shin- @jfrazelle There's a bit of a race in lxc-driver when determining the exitcode. When the driver determines that an lxc container has started, it immediately notifies the callback channel that the container has started. Right after this, on another channel, the error from the container comes, but since the method call has returned, this error code never makes it back. This only affects the daemon mode(-d) on docker run and start.

File: integration-cli/requirements.go (hunk 3 above)
Comment (tianon) [inline integration-cli/requirements.go]: Lol email ordering :+1:

Comment (jfrazelle) [inline integration-cli/requirements.go]: updated

File: integration-cli/requirements.go [hunk 4]
```diff
@@ -25,8 +30,37 @@ var (
...
+		func() bool {
+			return strings.HasPrefix(daemonExecDriver, "native")
+		},
+		"Test requires the native (libcontainer) exec driver.",
+	}
 )
 
+func init() {
+	// get daemon info
+	body, err := sockRequest("GET", "/info", nil)
```
Comment (ahmetalpbalkan) [inline integration-cli/requirements.go]: can we not do this lazy fetch inside the TestCondition func and cache there? when I run a single test case with `TESTFLAGS='-test.run XXX'` on my machine, I probably don't want to slow down of test bootstrap any longer.

Comment (jfrazelle) [inline integration-cli/requirements.go]: moved

**Comment by shin-** [comment]
Soo, that *should* work but for some reason lxc tests hang for me, sooo... docker/docker-py#509

**Comment by jfrazelle** [comment]
ping @icecrime @tiborvass

File: integration-cli/requirements.go [hunk 5]
```diff
@@ -25,6 +30,34 @@ var (
...
+	NativeExecDriver = TestRequirement{
+		func() bool {
+			if daemonExecDriver == "" {
+				// get daemon info
+				body, err := sockRequest("GET", "/info", nil)
+				if err != nil {
+					log.Fatalf("sockRequest failed for /info: %v", err)
+				}
+
+				var infoJSON map[string]interface{}
```
Comment (tiborvass) [inline integration-cli/requirements.go]: @jfrazelle I think it's better if you make it:

```Go
var infoJSON struct {
    ExecutionDriver string
}
```

That way, the Unmarshal will fail if it's not there, or if it's not a string.

**Comment by jfrazelle** [comment]
@tiborvass updated

**Comment by tiborvass** [comment]
LGTM

**Pull request merged**